src/Nix/String.hs

```diff
@@ -2,7 +2,7 @@
 {-# LANGUAGE OverloadedStrings #-}
 {-# OPTIONS_GHC -Wno-unused-top-binds #-}
 module Nix.String (
-    NixString
+    NixString(..)
```

Comment:
We were pretty intentionally not exporting this and I'm pretty hesitant to do so here. It should require explicit thought about context to handle strings internally. Is there a way to rewrite this PR without this? Maybe in terms of mappend, or possibly exporting some monad transformer here?

Reply:
Maybe add a comment there that it shouldn't be exported? :)

Reply:
I already encountered another place where I needed to get the context out and then construct a new NixString with a passed-in context.  That's the reason I created these functions:

https://github.com/haskell-nix/hnix/pull/405/files#diff-3a3776a3f4cdb0350fa1206cf782c14cR58

https://github.com/haskell-nix/hnix/pull/405/files#diff-3a3776a3f4cdb0350fa1206cf782c14cR140

At the end of the day there really aren't very many options.  Either you export the things that are needed by outside code or you add more complex functionality to this module that does all the things needed by outside functions.  That's king of what I was going for with this:

https://github.com/haskell-nix/hnix/pull/405/files#diff-3a3776a3f4cdb0350fa1206cf782c14cR143

Reply:
I guess this is ultimately a @jwiegley question... Do we want to guard against context misusage with specialized names/types or do we want to expect internal users to do the right thing and just export the constructors?

Reply:
I'd prefer using types to guard against misuse.